File: packages/shared/src/commonTest/kotlin/com/yancotv/shared/content/PlaylistDividersTest.kt
package com.yancotv.shared.content

import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class PlaylistDividersTest {

    @Test
    fun `real banners from the owner's account are dividers`() {
        // Every one of these is a row the provider ships as a heading, taken
        // verbatim from the live catalogue.
        val banners = listOf(
            "##### beIN SP⚽RTS ᴴᴰ #####",
            "###### RELAX ᵁᴴᴰ 3840P ######",
            "### ARABIC 24/7 4K UHD 3840P ###",
            "####### VIAPLAY PPV #######",
            "##### CY| ΚΥΠΡΙΑΚΑ/CYPRUS VIP #####",
            "#### DAZN EXCLUSIVE ᴴᴰ/ᴿᴬᵂ #####",
            "##### 24/7 SINGER #####",
            "##### ᵁᴴᴰ ³⁸⁴⁰ᴾ #####",
        )
        for (title in banners) assertTrue(isPlaylistDivider(title), title)
    }

    @Test
    fun `other separator characters count too`() {
        assertTrue(isPlaylistDivider("=== SPORTS ==="))
        assertTrue(isPlaylistDivider("***  MOVIES  ***"))
        assertTrue(isPlaylistDivider("▬▬▬ KIDS ▬▬▬"))
        assertTrue(isPlaylistDivider("   ### PADDED ###   "))
    }

    @Test
    fun `a row of nothing but the marker is a divider`() {
        assertTrue(isPlaylistDivider("##########"))
        assertTrue(isPlaylistDivider("------"))
    }

    @Test
    fun `channels people actually watch are not dividers`() {
        val real = listOf(
            "beIN SPORTS AFC 5",
            "V SPORT ᵁᴴᴰ ³⁸⁴⁰ᴾ",
            "Ping-Pong -- Live",
            "24/7 SINGER",
            "TRT 1 HD",
            "- Sky Sports",
            "E! Entertainment",
            "",
            "  ",
        )
        for (title in real) assertFalse(isPlaylistDivider(title), title)
    }

    @Test
    fun `both ends must have a run and of the same character`() {
        // A banner opened but not closed is a title with decoration, not a
        // heading — and mixing characters is not a convention any playlist
        // uses.
        assertFalse(isPlaylistDivider("### SPORTS"))
        assertFalse(isPlaylistDivider("SPORTS ###"))
        assertFalse(isPlaylistDivider("### SPORTS ==="))
        assertFalse(isPlaylistDivider("## SPORTS ##"))
    }

    @Test
    fun `a short run of the marker alone is not a divider`() {
        // "##" is two, under the floor, and "#" is a title someone could
        // plausibly have.
        assertFalse(isPlaylistDivider("##"))
        assertFalse(isPlaylistDivider("#"))
    }
}

File: tests/unit/playlist-dividers.test.ts
import { describe, it, expect } from 'vitest';
import { isPlaylistDivider } from '@yancotv/core';

/**
 * The rule is deliberately strict, and the tests exist to keep it that way.
 *
 * A looser rule starts eating real channels, which is a far worse failure than
 * leaving a banner in the list: a missing channel is invisible, while a banner
 * is merely annoying. Every "does NOT match" case below is a title a person
 * would actually want to watch.
 *
 * Kept in step with `PlaylistDividers.kt` — the same rule ships on Android and
 * iOS, and a change to what counts as a banner must not come to mean two
 * different things on two platforms.
 */
describe('isPlaylistDivider', () => {
  describe('matches real banners', () => {
    // Every one of these is from the owner's own catalogue.
    it.each([
      '##### 4K ᵁᴴᴰ ³⁸⁴⁰ᴾ #####',
      '###### RELAX ᵁᴴᴰ 3840P ######',
      '### ARABIC 24/7 4K UHD 3840P ###',
      '### beIN SP⚽RTS AFC 8K ###',
      '=== SPORTS ===',
      '--- MOVIES ---',
      '***  KIDS  ***',
    ])('%s', (title) => {
      expect(isPlaylistDivider(title)).toBe(true);
    });

    it('matches a row that is nothing but the marker', () => {
      expect(isPlaylistDivider('######')).toBe(true);
      expect(isPlaylistDivider('=========')).toBe(true);
    });

    it('ignores surrounding whitespace', () => {
      expect(isPlaylistDivider('   ### NEWS ###   ')).toBe(true);
    });
  });

  describe('does NOT match real channel names', () => {
    /**
     * The load-bearing half. A false positive here deletes a channel from the
     * user's library with no message and no way to find out why.
     */
    it.each([
      'beIN SPORTS 1 HD',
      'Ping-Pong -- Live',      // has a run, but not at both ends
      '24/7 SINGER',            // digits and a slash, no marker run
      '- MBC 1',                // one leading dash is not a heading
      'MBC 1 -',                // one trailing dash either
      'AR | BEIN SPORTS',       // pipe separator, not a run
      'C++ Programming TV',     // internal run, not at the ends
      '2001: A Space Odyssey',
      '*',                      // too short to be a run
      '**',
      '##',                     // two is below the three-character minimum
      '',
    ])('%s', (title) => {
      expect(isPlaylistDivider(title)).toBe(false);
    });

    /**
     * Both ends must use the SAME character. A title that happens to start with
     * one marker and end with another is not a banner — and this is the case a
     * naive "starts or ends with punctuation" rule gets wrong.
     */
    it('requires the same marker at both ends', () => {
      expect(isPlaylistDivider('### NEWS ===')).toBe(false);
      expect(isPlaylistDivider('--- NEWS ***')).toBe(false);
    });

    it('requires three or more at BOTH ends, not just one', () => {
      expect(isPlaylistDivider('### NEWS #')).toBe(false);
      expect(isPlaylistDivider('# NEWS ###')).toBe(false);
    });
  });

  it('survives null and undefined rather than throwing', () => {
    // Provider data is not validated by anyone before it reaches this.
    expect(isPlaylistDivider(undefined as unknown as string)).toBe(false);
    expect(isPlaylistDivider(null as unknown as string)).toBe(false);
  });
});
